File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int T = Integer.parseInt(br.readLine());

        for (int i = 0; i < T; i++) {
            Stack<String> stack = new Stack<>();
            String[] s = br.readLine().split("");
            String ans = "YES";

            for(String x : s) {
                if (x.equals("(")) {
                    stack.push(x);
                } else {
                    if(!stack.isEmpty()) {
                        stack.pop();
                    }
                    else {
                        ans = "NO";
                        break;
                    }
                }
            }
            if (!stack.isEmpty()) ans = "NO";
            bw.write(ans + "\n");
        }
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/9012. 괄호/괄호.kt
import java.util.Stack

fun main(args: Array<String>) {
    val size = readLine()!!.toInt()
    for (i in 0 until size) {
        val line = readLine()
        val stack = Stack<Char>()
        var ans = "YES"
        line?.let {
            for (c in line) {
                if (c == '(') stack.push(c)
                else {
                    if (stack.size > 0 && stack.peek() == '(') stack.pop()
                    else ans = "NO"
                }
            }
        }
        if (stack.size > 0) ans = "NO"
        println(ans)
    }
}